src/Core/Module/ModuleRepository.php

```diff
@@ -0,0 +1,249 @@
+    {
+        $moduleName = $event->getModule()->get('name');
+        if ($this->cacheProvider->contains($moduleName)) {
+            $this->cacheProvider->delete($moduleName);
+        }
+    }
+
+    public function getList(): array
+    {
+        $modules = [];
+        $modulesDirsList = $this->finder->directories()
+            ->in($this->modulePath)
+            ->depth('== 0')
+            ->exclude(['__MACOSX'])
+            ->ignoreVCS(true);
+
+        foreach ($modulesDirsList as $moduleDir) {
+            $moduleName = $moduleDir->getFilename();
+            if (!file_exists($this->modulePath . $moduleName . '/' . $moduleName . '.php')) {
+                continue;
+            }
+
+            $modules[] = $this->getModule($moduleName);
+        }
+
+        return $this->mergeWithModulesFromHook($modules);
+    }
+
+    public function getInstalledModules(): array
+    {
+        return array_filter($this->getList(), function ($module) {
+            return $module->database->get('installed') === 1;
+        });
+    }
+
+    public function getConfigurableModules(): array
+    {
+        return array_filter($this->getList(), function (Module $module) {
+            return $module->hasValidInstance() && !empty($module->getInstance()->warning);
+        });
+    }
+
+    public function getUpgradableModules(): array
+    {
+        return array_filter($this->getList(), function (Module $module) {
+            return $module->canBeUpgraded();
+        });
+    }
+
+    /**
+     * @return Module
+     */
+    public function getModule(string $moduleName): ModuleInterface
+    {
+        $path = $this->modulePath . '/' . $moduleName;
+        $filePath = $path . '/' . $moduleName . '.php';
+
+        $filemtime = (int) @filemtime($filePath);
+
+        if ($this->cacheProvider->contains($moduleName)) {
+            $module = $this->cacheProvider->fetch($moduleName);
+            if ($module->disk->get('filemtime') === $filemtime) {
+                return $module;
+            }
+        }
+
+        $isValid = $filemtime > 0 && $this->moduleDataProvider->isModuleMainClassValid($moduleName);
+        $attributes = $this->getModuleAttributes($moduleName, $isValid);
+        $disk = $this->getModuleDiskAttributes($moduleName, $isValid, $filemtime);
+        $database = $this->getModuleDatabaseAttributes($moduleName);
+
+        $this->cacheProvider->save($moduleName, new Module($attributes, $disk, $database));
+
+        return $this->cacheProvider->fetch($moduleName);
+    }
+
+    public function getModulePath(string $moduleName): ?string
+    {
+        $path = $this->modulePath . '/' . $moduleName;
+        $filePath = $path . '/' . $moduleName . '.php';
+
+        if (!is_dir($path) || !is_file($filePath)) {
+            return null;
+        }
+
+        return $path;
+    }
+
+    public function generateActionUrls(ModuleCollection $collection): ModuleCollection
```

Comment:
Method name is a bit confusing. Reader might assume it is supposed to return an array of URL strings.